qiskit/primitives/containers/bit_array.py

```diff
@@ -0,0 +1,337 @@
+
+    def _get_counts(
+        self, *, loc: int | Tuple[int, ...] | None, converter: Callable
+    ) -> Dict[str | int, int]:
+        if loc is None and self.size == 1:
+            loc = (0,) * self.ndim
+
+        elif loc is None:
+            raise ValueError(
+                f"Your BitArray has shape {self.shape}, meaning that it actually represents "
+                f"{self.size} different count dictionaries. You need to use the `loc` argument of "
+                "this function to pick one of them."
+            )
+
+        counts = defaultdict(int)
+        for shot_row in self._array[loc]:
+            counts[converter(shot_row.tobytes())] += 1
+        return dict(counts)
+
+    def bitcount(self) -> NDArray[np.uint64]:
+        """Compute the number of ones appearing in the binary representation of each shot.
+
+        Returns:
+            A ``numpy.uint64``-array with shape ``(*shape, num_shots)``.
+        """
+        return _WEIGHT_LOOKUP[self._array].sum(axis=-1)
+
+    @staticmethod
+    def from_bool_array(
+        array: NDArray[np.bool_], order: Literal["big", "little"] = "big"
+    ) -> "BitArray":
+        """Construct a new bit array from an array of bools.
+
+        Args:
+            array: The array to convert, with "bitstrings" along the last axis.
+            order: One of ``"big"`` or ``"little"``, indicating whether ``array[..., 0]``
+                correspond to the most significant bits or the least significant bits of each
+                bitstring, respectively.
+
+        Returns:
+            A new bit array.
+        """
+        array = np.array(array, dtype=bool, copy=False)
+
+        if array.ndim < 2:
+            raise ValueError("Expecting at least two dimensions.")
+
+        if order == "little":
+            # np.unpackbits assumes "big"
+            array = array[..., ::-1]
+
+        num_bits = array.shape[-1]
+        if remainder := (-num_bits) % 8:
+            # unpackbits pads with zeros on the wrong side with respect to what we want, so
+            # we manually pad to the nearest byte
+            pad = np.zeros(shape_tuple(array.shape[:-1], remainder), dtype=bool)
+            array = np.concatenate([pad, array], axis=-1)
+
+        return BitArray(np.packbits(array, axis=-1), num_bits=num_bits)
+
+    @staticmethod
+    def from_counts(
+        counts: Mapping[str | int, int] | Iterable[Mapping[str | int, int]],
+        num_bits: int | None = None,
+    ) -> "BitArray":
+        """Construct a new bit array from one or more ``Counts``-like objects.
+
+        The ``counts`` can have keys that are (uniformly) integers, hexstrings, or bitstrings.
+        Their values represent numbers of occurrences of that value.
+
+        Args:
+            counts: One or more counts-like mappings.
+            num_bits: The desired number of bits per shot. If unset, the biggest value found sets
+                this value.
+
+        Returns:
+            A new bit array.
```

Comment:
Should this say the shape? eg
```suggestion
            A new bit array with shape `()` for single input counts, or `(N,)` for an iterable of N counts.
```